source/client_channel_handler.c

```diff
@@ -41,6 +41,27 @@ static int s_packet_handler_default(
     return aws_raise_error(AWS_ERROR_MQTT_INVALID_PACKET_TYPE);
 }
 
+static void s_on_time_to_ping(struct aws_channel_task *channel_task, void *arg, enum aws_task_status status) {
+    (void)channel_task;
+
+    if (status == AWS_TASK_STATUS_RUN_READY) {
+        struct aws_mqtt_client_connection *connection = arg;
+        aws_mqtt_client_connection_ping(connection);
+        aws_channel_task_init(&connection->ping_task, s_on_time_to_ping, connection);
+
+        uint64_t schedule_time = 0;
+        aws_channel_current_clock_time(connection->slot->channel, &schedule_time);
+        if (connection->keep_alive_time) {
+            schedule_time +=
```

Comment:
Should we try to reduce the task time so that the ping is sent just under the timeout value, so that there's no chance of latency causing a hangup?

Reply:
I subtracted the request timeout out of the to run time. The only downside here is if someone does something dumb and pings every second with a 3 second timeout, they'll be running pings all day.

Reply:
but now that TCP keep alive can successfully pick up dropped connections within a second, we can just advise people to let this be set at the max value.

Reply:
How about we just keep using the one "keepalive" value, instead of introducing a second value. Here's what the MQTT spec says about keepalive on the server-side:
> If the Keep Alive value is non-zero and the Server does not receive a Control Packet from the Client within one and a half times the Keep Alive time period, it MUST disconnect the Network Connection to the Client as if the network had failed [MQTT-3.1.2-24].

We could follow that approach too. If we send a PINGREQ every 10 seconds, it should take 15 seconds without a PINGRESP before we kill the connection. It will confuse users that keepalive 10s means it takes 15s to timeout, but we could just document that. Or we could let user-passed-keepalive of 10s translate to pings every 6.66s, but that's not "literal mqtt keepalive" behavior. Choose the form of your confusor I guess

Reply:
Actually, the spec says that ANY packet type is sufficient to keep the connection alive, not just pings. As an optimization would could only send pings when no other kind of packet has been sent in keepalive seconds

Reply:
> It is the responsibility of the Client to ensure that the interval between Control Packets being sent does not exceed the Keep Alive value. In the absence of sending any other Control Packets, the Client MUST send a PINGREQ Packet

Reply:
I don't think this one is relevant anymore, closing.